staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go

```diff
@@ -660,6 +697,170 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
+	// is nothing to do here.
+	if owners := csaupgrade.FindFieldOwners(
+		accessor.GetManagedFields(),
+		metav1.ManagedFieldsOperationApply,
+		lastAppliedAnnotationFieldPath,
+	); len(owners) > 0 {
+		return nil
+	}
+
+	justAnnotation := &unstructured.Unstructured{}
+	justAnnotation.SetGroupVersionKind(info.Mapping.GroupVersionKind)
+	justAnnotation.SetName(accessor.GetName())
+	justAnnotation.SetNamespace(accessor.GetNamespace())
+	justAnnotation.SetAnnotations(map[string]string{
+		corev1.LastAppliedConfigAnnotation: originalAnnotation,
+	})
+
+	modified, err := runtime.Encode(unstructured.UnstructuredJSONScheme, justAnnotation)
+	if err != nil {
+		return nil
+	}
+
+	newObj, err := helper.Patch(
+		info.Namespace,
+		info.Name,
+		types.ApplyPatchType,
+		modified,
+		&metav1.PatchOptions{
+			FieldManager: fieldManager,
+		})
+	if err != nil {
+		return err
+	}
+
+	return info.Refresh(newObj, false)
+}
+
+// Check if the returned object needs to have its kubectl-client-side-apply
+// managed fields migrated server-side-apply.
+//
+// field ownership metadata is stored in three places:
+//   - server-side managed fields
+//   - client-side managed fields
+//   - and the last_applied_configuration annotation.
+//
+// The migration merges the client-side-managed fields into the
+// server-side-managed fields, leaving the last_applied_configuration
+// annotation in place (since SSA will keep it up-to-date)
+func (o *ApplyOptions) migrateToSSAIfNecessary(
+	helper *resource.Helper,
+	info *resource.Info,
+) error {
+	accessor, err := meta.Accessor(info.Object)
+	if err != nil {
+		return err
+	}
+
+	// To determine which field managers were used by kubectl for client-side-apply
+	// we search for a manager used in `Update` operations which owns the
+	// last-applied-annotation.
+	//
+	// This is the last client-side-apply manager which changed the field.
+	//
+	// There may be multiple owners if multiple managers wrote the same exact
+	// configuration. In this case there are multiple owners, we want to migrate
+	// them all.
+	csaManagers := csaupgrade.FindFieldOwners(
+		accessor.GetManagedFields(),
+		metav1.ManagedFieldsOperationUpdate,
+		lastAppliedAnnotationFieldPath)
+
+	// Migrate each csa manager which claims to own the last applied annotation
+	for _, csaManagerName := range csaManagers {
+		// Re-attempt patch as many times as it is conflicting due to ResourceVersion
+		// test failing
+		var err error
+		for i := 0; i < maxPatchRetry; i++ {
+			var patchData []byte
+			var obj runtime.Object
+
+			patchData, err = csaupgrade.UpgradeManagedFieldsPatch(
+				info.Object, csaManagerName, o.FieldManager)
+
+			if err != nil {
+				// If patch generation failed there was likely a bug.
+				return err
+			} else if patchData == nil {
+				// nil patch data means nothing to do - object is already migrated
+				return nil
+			}
+
+			// Send the patch to upgrade the managed fields if it is non-nil
+			obj, err = helper.Patch(
```

Comment:
read my mind. Just was working on a change to do this as a single PATCH call rather than the complicated loop logic. 

But I would expect this case to be pretty rare anyway.